Hazel/src/Hazel/ImGui/ImGuiConsole.cpp

```diff
@@ -0,0 +1,249 @@
+#include "hzpch.h"
+#include "ImGuiConsole.h"
+
+#include "imgui.h"
+#include "examples/imgui_impl_glfw.h"
+#include "examples/imgui_impl_opengl3.h"
+
+namespace Hazel {
+
+	float ImGuiConsole::s_DisplayScale = 1.0f;
+	uint16_t ImGuiConsole::s_MessageBufferCapacity = 200;
+	uint16_t ImGuiConsole::s_MessageBufferSize = 0;
+	uint16_t ImGuiConsole::s_MessageBufferBegin = 0;
+	ImGuiConsole::Message::Level ImGuiConsole::s_MessageBufferRenderFilter = ImGuiConsole::Message::Level::Trace;
+	std::vector<Ref<ImGuiConsole::Message>> ImGuiConsole::s_MessageBuffer(s_MessageBufferCapacity);
+	bool ImGuiConsole::s_AllowScrollingToBottom = true;
+	bool ImGuiConsole::s_RequestScrollToBottom = false;
+
+	void ImGuiConsole::AddMessage(Ref<Message> message)
```

Comment:
Why make this a ref which has shared ownership? It should be able to just be a message, if you construct the message or move it into here, then s_Message buffer can be a vector of Messages and you avoid additional indirection and unclear ownership of messages